### PR TITLE
ES 2015

### PR DESCRIPTION
**Pull request by jonathanong**

### CODE REVIEW

**Comment by dead-horse** [comment]
LGTM

**Comment by tejasmanohar** [comment]
thoughts on es6 classes? requires a major API change- i.e. you _must_ use the `new` keyword afaik

```
var app = koa();
/// becomes
var app = new koa();
```

(i think, that's one of the ES2015 features I haven't used much)

classes don't really offer any benefits - 100% sugar - questioning whether it's a good refactor or not.

**Comment by felixfbecker** [comment]
Not related directly but any comments on why koa exports prototypes in source code? I find that very confusing.

File: lib/request.js [hunk 1]
```diff
@@ -3,12 +3,12 @@
...
  */
 
-var contentType = require('content-type');
-var stringify = require('url').format;
-var parse = require('parseurl');
-var qs = require('querystring');
-var typeis = require('type-is');
-var fresh = require('fresh');
+const contentType = require('content-type');
+const stringify = require('url').format;
```
Comment (ruimarinho) [inline lib/request.js]: Can't this be deprecated in favor with es6 template strings?

Comment (tejasmanohar) [inline lib/request.js]: @ruimarinho can't exactly tell what you're pointing to here. which line/statement/dependency?

**Comment by tejasmanohar** [comment]
@felixfbecker hm. in which context?

**Comment by tejasmanohar** [comment]
another thing to note- `class Application extends Emitter` would (roughly) translate to `Object.setPrototpyeOf(Application, Emitter)` rather than `Object.setPrototypeOf(Application.prototype, Emitter.prototype);` -> https://github.com/koajs/koa/blob/master/lib/application.js#L56

**Comment by felixfbecker** [comment]
in context.js for example

**Comment by tj** [comment]
@felixfbecker nothing special, just separates the protos into different files, they're used here https://github.com/koajs/koa/blob/master/lib/application.js#L47

**Comment by felixfbecker** [comment]
it just confused me when I wanted to unit test middleware and tried to somehow create a context manually to call the middleware with. Could not get it to work.

**Comment by tejasmanohar** [comment]
@felixfbecker ah, hmm... not sure what the problem would be there. maybe open another issue w/ isolated, runnable example?

File: lib/request.js (hunk 1 above)
Comment (jonathanong) [inline lib/request.js]: `stringify = require('url').format`? not sure how that works

**Comment by jonathanong** [comment]
LOL so many merge conflicts now

**Comment by tejasmanohar** [comment]
aghh!

**Comment by tejasmanohar** [comment]
someone needs to write a regex / pattern to solve these.... lol

**Comment by juliangruber** [comment]
ETOOMANYCONFLICTS

**Comment by tejasmanohar** [comment]
@jonathanong how'd you solve the conflicts so fast? :tada:

**Comment by tejasmanohar** [comment]
@jonathanong Is this basically our v2 branch? i.e. Should all changes to be reviewed for v2 be sent here (else, where? master isn't ideal.)?

**Comment by jonathanong** [comment]
@tejasmanohar yeah for now. i think branching off `v1.x` from `master` and making this master would be a good idea too.

**Comment by tejasmanohar** [comment]
@jonathanong aight.

>i think branching off v1.x from master and making this master would be a good idea too.
i agree

**Comment by jonathanong** [comment]
merging into master!

**Comment by tejasmanohar** [comment]
alright, then classes will come in a separate PR to master when i get some free time :)

**Comment by tejasmanohar** [comment]
@jonathanong is there a reason we have a separate branch for v1 rather than just a git tag? https://github.com/koajs/koa/tree/v1.x.

**Comment by jonathanong** [comment]
@tejasmanohar in case we need to maintain it more

**Comment by tejasmanohar** [comment]
@jonathanong aha makes sense

**Pull request closed without merging**